Separates Kaggle's modified BigQuery client into its own library and expose a custom Credentials for GCP integration.

vimota opened the pull request:
This change separates the current code modifying the BigQuery client into its own library that the user can directly import and use - but to maintain backwards compatibility we still monkeypatch the client to inject this library. This also adds a custom Credential class that we will use to refresh the user's oauth token ourselves using the account they've connected with the Kernel.
Context: go/kaggle-oauth-kernels

rosbo commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
+    bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+    specified_project = kwargs.get('project').lower() if 'project' in kwargs else None
```
For the existing use case, `specified_project` will be set to `None` so it will enter the else block in 41. This means, the existing use case is broken right? Am I missing something?

vimota replied:
If specified_project is None AND kaggle_proxy_token is None (which should be the case, for the existing use case, since that will only be populated for Kernels where the user has connected an account) then we should still use the proxy.

It's a little confusing but it should be something like this:
 if (kernel has no connected gcp account, ie. no jwt present):
      -  if (user specified a project that's not 'kaggle') : throw an error, since that won't work anyways (today this also already fails)
      - else if (user specified 'kaggle' project or no project): existing behaviour, use data proxy.
if (kernel does have connected gcp account):
       - if (use specifies 'kaggle' project): they are indicating they want to use the proxy explicitly.
        - else : proceed using their connected account instead of proxy (whether or not they specified a proxy).

rosbo replied:
Currently, `KAGGLE_DATA_PROXY_TOKEN` is set to a value to allow access to public BQ as far as I know:
https://www.kaggle.com/rosebv/kaggle-data-proxy-token?scriptVersionId=8721257

This means `kaggle_proxy_token` will be set right?

vimota replied:
Sorry I meant : If specified_project is None AND *bq_user_jwt* is None (which should be the case, for the existing use case, since that will only be populated for Kernels where the user has connected an account) then we should still use the proxy.

Yeah, kaggle_proxy_token  should always be set.

rosbo replied:
👍

rosbo commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
+kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
+CONNECTION_BASE_URL = Connection.API_BASE_URL
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
```
Consider passing `data_proxy_project` and `bq_user_jwt` in the kwargs of this method. At line 55, just set the default values for these parameters with `os.getenv(...)`. 

This way, in the tests, you can override them without having to mess with the environment variable.

vimota replied:
Do you mean, then I would pass those in the bigquery client creation in the test? Ie. client = bigquery.Client(project='KAGGLE', bq_user_jwt=foo)? The only reason I don't like that is that it modifies the interface of the bigquery client library just for testing purposes. Let me see if I can come up with something else that works otherwiwse I'll do this :) 
By the way, would there be an issue with change the env var in the tests (and then clean them up to what they were)? In particular, if tests run in parallel I could see that being an issue. I almost want to have a integration style test for this where it does a separate docker run for each test since it'll be a clean environment each time and any modified global state of code from sitecustomize.py from one test won't affect the other.

rosbo replied:
I am mostly concerned about the fact that we run our tests in parallel.

rosbo replied:
Are you planning to address this in a separate PR or this PR? (either way is fine with me).

vimota replied:
Just did it in the latest commit :) Learned something new today: EnvironmentVarGuard is pretty handy!

sebbov replied:
fyi tests don't run in parallel (unless something changed).  Still 👍 on the change though.

vimota commented:
Friendly ping :)

sebbov commented:
Sorry for the delay, still want to review this before you merge.  Reviewing now.

sebbov commented:
Victor, can you provide more context on how this fits into the big picture.  E.g. link to design doc in the PR description.  Not just for my review, also for future blaming / understanding.

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
 # TODO(rosbo): Remove this once we fix the issue with fastai importing older libcudnn if imported prior to tensorflow
 import tensorflow
 import os
+from google.auth import credentials
```
These imports are no longer conditional on `KAGGLE_DATA_PROXY_TOKEN` being set.  This could be detrimental for non-Kaggle kernels use of our Docker image.  I suppose only startup latency of any Python binary, but there could be more to it than that.  If you do plan for it, suggest to blame back and find from the author why that approach was taken (hopefully that doesn't circle back to me ;) ).

sebbov replied:
ping

vimota replied:
Sorry I hadn't finished addressing everything yet :) I wish Github PRs were more like critique and allowed setting the attention set only once you're done.

vimota replied:
I can make all of this conditional on (KAGGLE_DATA_PROXY_TOKEN || KAGGLE_BQ_USER_JWT) ?

sebbov replied:
SGTM

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
 # TODO(rosbo): Remove this once we fix the issue with fastai importing older libcudnn if imported prior to tensorflow
 import tensorflow
 import os
+from google.auth import credentials
```
nit:  Imports in alpha order.

The `tensorflow` import is an exception to this:  The point is to always import it prior to any other imports.

So:

```
# TODO(rosbo)...
import tensorflow

from google.auth import bigquery
...
import os
```

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
-    from google.cloud import bigquery
-    from google.cloud.bigquery._http import Connection
+class KaggleKernelCredentials(credentials.Credentials):
+    def refresh(self, request):
```
nit: newline before method def.

sebbov replied:
ping

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
+            credentials=anon_credentials,
+            **kwargs)
+    else:
+        Connection.API_BASE_URL = CONNECTION_BASE_URL
```
Not sure why you need this and `CONNECTION_BASE_URL` at all.  Why not just leave it to what it is in the first place?

vimota replied:
My thinking was that, the first time this runs on startup, we store `CONNECTION_BASE_URL = Connection.API_BASE_URL` as the default url that the bigquery library wants to use. And then each time the user uses the bigquery client library, we will update Connection.API_BASE_URL to either the data proxy url (line 32) or back to the default url (line 41) depending on what project the user uses (Ie. if user indicates they want to use the kaggle proxy or their own account). 

Setting it this way, is only to prevent issues if the user first instantiates the client for the kaggle proxy and then switches back to using their own project or vice versa.

sebbov replied:
> Setting it this way, is only to prevent issues if the user first instantiates the client for the kaggle proxy and then 
> switches back to using their own project or vice versa.

Can we actually support that?  This is a static class variable we're changing here right?  Is an existing client guaranteed not to reuse this?

vimota replied:
That's a good point. This will work if the user creates clientA (kaggle proxy) and uses it then creates clientB (non proxy) and uses it, but won't work if they interleave their usage. It's a little less broken but still broken - do you have any ideas on how to get this to work or should we disallow?

One way I can imagine us disallowing it is checking if the current client making a call is of the same type as the connection (ie. a project=kaggle client call should ensure that the Connection is the one expected otherwise throw an error). Now that I think of it, we can probably intercept each client call and update the Connection.API_BASE_URL at the time of the call so that it's always the correct value. How does that sound?

sebbov replied:
The right solution is not to do something that isn't supported :).  The existing approach is already a hack since it accesses internals and modifies a static that I don't believe is meant to be changed.

So, the right solution is to modify upstream to support this all properly.  IIRC this was my original suggestion already.  Monkey patching is only at most okay in the short term.

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
+    bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+    specified_project = kwargs.get('project').lower() if 'project' in kwargs else None
```
`kwargs['project'].lower() if 'project' in kwargs else None`
may be a bit more clear as `.get()` covers the case you already tested for.  One no longer has to worry about the case that `.get()` would return the default `None`.

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
+    if bq_user_jwt is None and specified_project and specified_project != 'kaggle':
+        raise Exception("In order to query a private BigQuery project, please connect a GCP account. "
+                        "Otherwise specify 'kaggle' as the project to use Kaggle's public dataset BigQuery integration.")
+    use_data_proxy = specified_project == 'kaggle' or bq_user_jwt is None
```
Should `'kaggle'` be `'kaggle-161607'`?
  https://github.com/Kaggle/Kaggle.SDS.Worker/blob/720f4aa8ad15549e2c1069c557ccca2480dd9f32/kernels/worker/src/kernel_run_container_commands.py#L62

I'm not sure why you need that test in the first place.  It doesn't exist on the LHS for this case, and it seems preferable not to specify the kaggle project again, nor to expose it more than it already is.

vimota replied:
This kaggle "project" the user specifies is really more of just a way to tell us they want to use the Kaggle "free bigquery" Data Proxy even though they've connected a GCP account. If they haven't connected their GCP account (ie. bq_user_jwt is not present) then they don't need specify a 'kaggle' or any other project, we will default to using the data proxy. 

More info here: https://github.com/Kaggle/docker-python/pull/414#discussion_r243672377

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
+        if data_proxy_project is None or kaggle_proxy_token is None:
+            # We don't have the data proxy info so leave the bq client unmodified.
+            return bq_client(*args, **kwargs)
+        print("Using Kaggle's public dataset BigQuery integration.")
```
This output would be new.  Are we sure it is desirable?  I suppose this goes in the cell output?  If so, is it only done when a bq dataset is attached to the kernel?

vimota replied:
I mentioned to others that I was thinking of adding this messaging  (https://kaggle.slack.com/archives/CDB736R0C/p1545422371002800?thread_ts=1545422009.001500&cid=CDB736R0C) and they seemed to like  the idea. I don't think a BQ dataset being attached to the Kernel is related here, but correct me if I'm wrong - it should only display when the user instantiates a BigQuery client (ie. client = bigquery.Client(project='KAGGLE'))

sebbov replied:
Ah yes, that seems fine then indeed.  Though IMO the intent should be expressed by the user in the first place, not some magic behind the scenes and then telling the user what the magic did.

vimota replied:
I agree I'd like the least amount of magic as possible. However we already have some magic going on (Kaggle data proxy injecting our own credentials) and I figured now that we're adding another - less magical way of using it - the user should know which one is being used. My hope is that the user explicitly "tells" us which way they want to use (ie. project = 'mypersonalproject' or 'kaggle') but this only makes it more clear to them what is happening.

sebbov replied:
There is an opportunity to avoid the magic (and hopefully evt get to a place without any magic) here, let's take it.  Project 'kaggle' isn't a real GCP product (or if it is wouldn't be used here).

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
+            **kwargs)
+    else:
+        Connection.API_BASE_URL = CONNECTION_BASE_URL
+        Connection._EXTRA_HEADERS.pop('X-KAGGLE-PROXY-DATA', None)
```
How would this header ever be present?

vimota replied:
In the same case I described above, if the user instantiated a bigquery client first using the Kaggle proxy (ie. client = bigquery.Client(project='KAGGLE')) and then recreated the client with their own private project. It should not be a common case, but wanted to prevent breakage if they did this. Perhaps I should add a unit test for this case.

sebbov replied:
Resolving this since we're covering it above.

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
+            return bq_client(*args, **kwargs)
+        print("Using enabled BigQuery integration.")
+        kwargs['credentials'] = KaggleKernelCredentials()
+        kwargs['project'] = kwargs.get('project')
```
Seems like this is a nop unless `project` is not already there and you want to explicitly set to `None`?  If so, I'm not sure why it is needed (isn't that the default kwargs value in the function definition already)?

vimota replied:
Removed.

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
+    bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+    specified_project = kwargs.get('project').lower() if 'project' in kwargs else None
```
Not fully clear if you need the `.lower()` here.

vimota replied:
In case the user passes "Kaggle" or "KAGGLE" etc in bigquery.client(), but you're probably right - I'll move the lower() to just the comparison condition checking.

sebbov commented on tests/test_bigquery.py:
```diff
@@ -1,38 +1,67 @@
-        threading.Thread(target=httpd.serve_forever).start()
-        client = bigquery.Client()
+    def _test_proxy(self, client, should_use_proxy):
+        class HTTPHandler(BaseHTTPRequestHandler):
```
Can this be left at the module-level for better readability?  I don't see any currying taking place but I might be missing it.

vimota replied:
Yeah, I spent some time trying to do this properly but the issue is that HTTPHandler.called is a class variable, so if it is at the module level, the first time it's set to true from one test running, other tests will see it as true (in fact this was happening in my tests at first). Ideally 'called' would be an instance variable but HTTPServer/TCPServer never actually keep around the instance of HTTPHandler (they call RequestHandlerClass internally) for us to be able to inspect. So this is kind of a hack to keep the class variable scoped on a per test basis.

sebbov replied:
You could reset the class attribute in the test's setup method, or here?  Tests don't run in parallel and never will.

You can also define a base class module-level, then subclass that within this function with `called = False`.

sebbov replied:
Yeah what you do is fine if you prefer it.  Tests always run sequentially in google3 (parallelism is provided at a coarser level, not in-process), which is why I'd prefer the above for clarity / avoiding too much nesting.  Sorry for the subclass suggestion, that won't work indeed.

Having a fake BQ server would probably help too, but none of that is provided for GCP APIs.

sebbov commented on tests/test_bigquery.py:
```diff
@@ -1,38 +1,67 @@
-    header_found = False
+            httpd.shutdown()
+            if should_use_proxy:
+                self.assertTrue(HTTPHandler.called, msg="Fake server did not recieve a request from the BQ client.")
```
`receive`

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,22 +1,56 @@
-    from google.auth import credentials
-    from google.cloud import bigquery
-    from google.cloud.bigquery._http import Connection
+class KaggleKernelCredentials(credentials.Credentials):
```
ping

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,20 +1,56 @@
+    if bq_user_jwt is None and specified_project and specified_project.lower() != 'kaggle':
+        raise Exception("In order to query a private BigQuery project, please connect a GCP account. "
+                        "Otherwise specify 'kaggle' as the project to use Kaggle's public dataset BigQuery integration.")
+    use_data_proxy = (specified_project and specified_project.lower() == 'kaggle') or bq_user_jwt is None
```
Continued from:

```
@vimota
vimota 3 days ago  Member
This kaggle "project" the user specifies is really more of just a way to tell us they want to use the Kaggle "free bigquery" Data Proxy even though they've connected a GCP account. If they haven't connected their GCP account (ie. bq_user_jwt is not present) then they don't need specify a 'kaggle' or any other project, we will default to using the data proxy.

More info here: #414 (comment)
```

since github won't let me comment on that.

I find this really confusing.  That comment also specifies itself that it is confusing.  Don't explain confusing things in a review comment.  Document them to start with, including their motivation.

In this case I would like to understand why we're so hard trying to hijack the standard bq client to fit all sorts of customizations.  At least the "I explicitly want to use the Kaggle bq" could be more clearly supported that way.

vimota replied:
I had originally preferred doing something like that as well - ie. from Kaggle import bq (kaggle proxy) vs normal import bq (using connected account), but opted for this because it avoided breaking users existing Kernels. Do you think it's better to take that approach even if it breaks existing users, I agree it'll be cleaner?

harrisse replied:
@sebbov And to provide a bit more context, the reason we want to keep everything as close as possible to the code people would write outside of kernels is because part of the appeal of GCP integrations from Google's perspective is teaching users how to use GCP tools. The fewer changes they need to make between Kaggle and non-Kaggle code, the easier they'll be able to use GCP for their non-Kaggle work. Or at least that's the thinking. I understand there's always going to be some magic involved wherever we want to do that, but I think BQ is likely the worst offender since there's both the concept of a public BQ product and private BQ product rolled into the same bq library.

sebbov replied:
@harrisse 👍 Does it sound right that this means:

- Let them use the vanilla BQ client for the non-proxy case. Provide an example on how to use the environment-specified credentials.
- For the proxy case, wrap this in a separate class (and note what you can't do in this case, e.g. the concurrent use w/ the vanilla client mentioned in a different comment) since there is no such thing as proxy support in the current BQ client at this time.

sebbov commented on tests/test_bigquery.py:
```diff
@@ -1,38 +1,74 @@
+                HTTPHandler.header_found = any(k for k in s.headers if k == "X-KAGGLE-PROXY-DATA" and s.headers[k] == "test-key")
+                s.send_response(200)
+
+        with HTTPServer((HOSTNAME, PORT), HTTPHandler) as httpd:
```
Where is this HTTP server used?  I don't see anything pointing to the address it serves.

vimota replied:
The server is started with the env var: KAGGLE_DATA_PROXY_URL=http://127.0.0.1:8000, so that's what the connection uses to make a call to this HTTP server.

sebbov replied:
Mhh okay :).  I can see the server serves `127.0.0.1:8000`, but not how `KAGGLE_DATA_PROXY_URL` is set to it (so the client would actually use it), or how we test the case where it isn't used by the client if that is still relevant.  I'm probably missing where this is done -- I am still struggling with PRs...

If `KAGGLE_DATA_PROXY_URL` is indeed set to `127.0.0.1:8000`, the server address (host/port components) used in this test should be built from it.  I know this is existing, but that avoid duplication 
(reducing the chance of breakages) and make this more readable.

sebbov commented on tests/test_bigquery.py:
```diff
@@ -1,38 +1,74 @@
 
 HOSTNAME = "127.0.0.1"
 PORT = 8000
 URL = "http://%s:%s" % (HOSTNAME, PORT)
```
Seems unused?  Remove if so, thanks.

vimota commented:
I believe I've addressed all the feedback and refactored it to take the approach we discussed where:
- We expose a kaggle_bq_client (feel free to suggest better name) which will be how users access the data proxy
- Users can use their own GCP credentials by using the vanilla bq with our KaggleKernelCredentials exposed in the Kaggle library (example in the unit tests).
- For now we preserve the monkeypatched data proxy (which essentially just exposes the library) until we update the documentation to use the library instead.

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,20 +1,26 @@
 kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
-if kaggle_proxy_token:
+bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+if (kaggle_proxy_token or bq_user_jwt):
```
Drop the `()`.

sebbov commented on patches/kaggle.py:
```diff
@@ -0,0 +1,40 @@
+
+
+class KaggleKernelCredentials(credentials.Credentials):
+        """Custom Credentials used to authenticate using the Kernel's connected OAuth account."""
```
Include sample usage.  Same for other public classes below.

While at it:  Could you update the PR title/description to match the new approach?

sebbov commented on patches/kaggle.py:
```diff
@@ -0,0 +1,40 @@
+
+
+class DataProxyConnection(Connection):
+        """Custom Connection class used to proxy the BigQuery client ho Kaggle's data proxy."""
```
s/ho/to/

sebbov commented on patches/kaggle.py:
```diff
@@ -0,0 +1,40 @@
+
+        def __init__(self, client):
+            super().__init__(client)
+            self._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
```
Leave a TODO to indicate that we shouldn't depend on the internals of the BQ client.  Ideally, file the issue to that client's project at this time and include the issue ref in the TODO comment.

sebbov commented on patches/kaggle.py:
```diff
@@ -0,0 +1,40 @@
+            self._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
+
+
+class DataProxyClient(bigquery.client.Client):
```
docstring

sebbov commented on patches/sitecustomize.py:
```diff
@@ -1,20 +1,26 @@
+            return bq_client(*args, **kwargs)
 
+    # Monkey patches BigQuery client creation to use proxy or user-connected GCP account.
+    # TODO: Remove monkeypatching altogether and move to using a Kaggle library for Data Proxy and
```
> ... using a Kaggle library ...

This is the library you introduce in this PR as well right?  If so, point specifically at it since it no longer is some future idea.

E.g.:

```
Deprecated in favor of Kaggle.DataProxyClient().
TODO: Remove this once uses have migrated to that new interface.
```

sebbov commented on tests/test_bigquery.py:
```diff
@@ -1,38 +1,86 @@
+                HTTPHandler.header_found = any(k for k in s.headers if k == "X-KAGGLE-PROXY-DATA" and s.headers[k] == "test-key")
+                s.send_response(200)
+
+        server_address = urlparse(os.getenv('KAGGLE_DATA_PROXY_URL'))
```
💯

sebbov commented on patches/kaggle.py:
```diff
@@ -0,0 +1,40 @@
+class DataProxyClient(bigquery.client.Client):
+    def __init__(self, project=None):
+        if project:
+            raise Exception("In order to query a private BigQuery project, please connect a GCP account. "
```
A private BigQuery project query shouldn't use this `kaggle.DataProxyClient` class in the first place?  If so, the exception message is misleading.

sebbov commented on patches/kaggle.py:
```diff
@@ -0,0 +1,40 @@
+        )
+        self._connection = DataProxyConnection(self)
+
+kaggle_bq_client = lambda *args, **kwargs:  DataProxyClient(*args, **kwargs)
```
I would remove this.  Instantiate where you need it instead (and only once).  Essentially, same as what the bigquery client lib provides.

The name also stutters (`kaggle.kaggle_...`) and is too wide of a scope (`bq_client` where it really is only for the data proxy case).

vimota replied:
What do you mean by 'Instantiate where you need it'? Expose it just as DataProxyClient? Sounds good to me, but do you think that name will confuse users - DataProxy is not really a term they're familiar with.

sebbov replied:
Sorry, it doesn't instantiate here, misread.  But essentially this is `kaggle_bq_client = DataProxyClient`, no point in the lambda?

Either way, what I meant is to expose the class (it actually already is), have users use that.  If the name is confusing / not what we want to expose, let's name it properly in the first place.  How about `kaggle.PublicBigqueryClient()`?

vimota replied:
Makes sense, that's what I figured you meant! I'll fix that in the next CL.

sebbov commented on Dockerfile:
```diff
@@ -505,6 +505,7 @@ RUN pip install --upgrade dask && \
 
 # Add BigQuery client proxy settings
 ENV PYTHONUSERBASE "/root/.local"
+ADD patches/kaggle.py /root/.local/lib/python3.6/site-packages/kaggle.py
```
I'm worried that we are going to see clashes with the Kaggle API, and perhaps with the competition support injected by the worker at container creation time.  Can you make sure this is fine with @rysteboe (for the former) and @dster2 (for the latter)?

dster2 replied:
Can't say I'm well-informed on what's going on here, but certainly we do need to keep support for the `kaggle.competitions.twosigmanews` module in Kernels, which is not installed as part of the public `kaggle` module.

vimota replied:
I can't find where kaggle/competitions/twosigmanews gets added to PYTHONPATH - I wonder if I can change how that gets done to so that it instead just copies competitions/ into an existing kaggle/ package folder created in this Dockerfile (which loads this kaggle.py file into it) that way everything is under the same namespace.
Alternatively maybe it is safer to just give it a different name? ie. from KaggleGCP import DataProxyClient?

vimota replied:
I investigated this a bunch today, and testing end to end I was able to see that indeed the namespaces collide and you lose kaggle.competitions.twosigmanews with this current approach. Finally I got this as a solution and tested it end to end: https://github.com/Kaggle/Kaggle.SDS.Worker/pull/543.

sebbov commented on patches/sitecustomize.py:
```diff
@@ -6,7 +6,7 @@
     from google.cloud import bigquery
     from google.cloud.bigquery._http import Connection
-    from kaggle import PublicBigqueryClient
+    from kaggle_gcp import PublicBigqueryClient
```
I suppose this should be `from kaggle.gcp import ...`, once that works (VMs are updated across the board), so to reflect the expected general usage?   => :Leave a TODO or make that change and hold until we're ready for it.

vimota replied:
We've discussed offline. The issue here is that the tests would fail if we did this since the worker is the one who imports kaggle_gcp into the kaggle.gcp namespace, and since that doesn't happen during the tests this would fail. I've added a comment todo to correct this.
